### ISSUE TITLE
Use mergeWithKey for Map and IntMap instances of Align

### ISSUE THREAD

**Issue by Zemyla** [isomorphism/these#48]
Instead of the current instances of Align for Map and IntMap, which require fmapping over each map and then using an unsafe unionWith to merge them, the mergeWithKey function should be used:

``` haskell
instance (Ord k) => Align (Map k) where
    nil = Map.empty
    alignWith fn = Map.mergeWithKey (\_ a b -> Just $ fn $ These a b) (fmap (fn . This)) (fmap (fn . That)) -- and similarly for IntMap
```

mergeWithKey is inlined after the three function arguments are given, thus providing opportunities for optimization; in addition, each subtree in the map is only traversed once instead of twice, and there's no chance of an `oops` happening.

**Comment by phadej** [comment]
Is this correct? Ot looks like that `fn` is called twice for common keys. Please make a PR, we should have tests for this IIRC.

**Comment by gelisam** [comment]
No, `fn` is only called once for common keys. Here is a test which sheds some light on what happens under the hood:

    {-# LANGUAGE OverloadedLists, RankNTypes #-}
    import Debug.Trace
    import Data.Map.Strict (Map)
    import Data.These
    import qualified Data.Map.Strict as Map
    import Text.Printf

    noisyApply :: Show a => String -> (a -> b) -> a -> b
    noisyApply showf f x = f $ trace (printf "%s (%s)" showf (show x)) x

    noisyFmap :: (Functor f, Show (f a)) => String -> (a -> b) -> f a -> f b
    noisyFmap showf f = noisyApply (printf "fmap %s" showf) (fmap f)

    -- |
    -- >>> test noisyApply noisyFmap Map.mergeWithKey
    -- fmap (fn . This) (fromList [])
    -- fn (These 2 4)
    -- fmap (fn . This) (fromList [])
    -- fmap (fn . This) (fromList [("left-only",1)])
    -- fmap (fn . That) (fromList [("right-only",3)])
    -- fn (This 1)
    -- fn (That 3)
    -- True
    test :: (String -> (These Int Int -> Int) -> These Int Int -> Int)
         -> (String -> (Int -> Int) -> Map String Int -> Map String Int)
         -> ( (String -> Int -> Int -> Maybe Int)
           -> (Map String Int -> Map String Int)
           -> (Map String Int -> Map String Int)
           -> Map String Int
           -> Map String Int
           -> Map String Int 
            )
         -> Bool
    test nap nmap mergeWithKey = alignWith (mergeThese (+)) [("left-only", 1), ("both",2)]
                                                            [("right-only",3), ("both",4)]
                              == [("left-only",1), ("right-only",3), ("both",6)]
      where
        alignWith fn = mergeWithKey (\_ a b -> Just $ nap "fn" fn $ These a b)
                                    (nmap "(fn . This)" (nap "fn" fn . This))
                                    (nmap "(fn . That)" (nap "fn" fn . That))

As you can see, `fmap (fn . This)` and `fmap (fn . That)` aren't called on the entire Map, instead they are called repeatedly on subsets of those Maps. Recall that a Map is represented as a binary tree. As [the documentation for `mergeWithKey`](http://hackage.haskell.org/package/containers-0.5.7.1/docs/Data-Map-Lazy.html#v:mergeWithKey) explains, the subtrees which are only present in the first map are passed" to `fmap (fn . This)`, and similarly for `fmap (fn . That)`.

You were probably thinking that `mergeWithKey` had the following implementation:

    -- |
    -- >>> test noisyApply noisyFmap badMergeWithKey
    -- fn (These 2 4)
    -- fmap (fn . This) (fromList [("both",2),("left-only",1)])
    -- fmap (fn . That) (fromList [("both",4),("right-only",3)])
    -- fn (This 1)
    -- fn (That 3)
    -- True
    badMergeWithKey :: Ord k
                    => (k -> a -> b -> Maybe c)
                    -> (Map k a -> Map k c)
                    -> (Map k b -> Map k c)
                    -> Map k a
                    -> Map k b
                    -> Map k c
    badMergeWithKey fxy fx fy xs ys = Map.mapMaybe id (Map.intersectionWithKey fxy xs ys)
                          `Map.union` fx xs
                          `Map.union` fy ys

That is, `fmap (fn . This)` and `fmap (fn . That)` are applied to the entire Maps, `fn` is applied to values in their intersection, and a left-biased union makes sure that the `These` outputs overwrite the `This` and `That` outputs where applicable. This approach would indeed traverse a bigger portion of the binary trees than necessary, but notice that `fn` is still applied to each value only once! This is because of laziness: even though I am using `Data.Map.Strict`, it is only the operations defined in that module which are strict, the datatype `Map` exported by that module is the same as the `Map` exported by `Data.Map.Lazy`. In particular, its `fmap` implementation is lazy, so if a `This` output gets overwritten by a `These` output, the `fn` call which produces the `This` output never gets forced.

Here's a variant with the strict `Data.Map.Strict.map`:

    -- |
    -- >>> test noisyApply noisyMap badMergeWithKey
    -- fn (These 2 4)
    -- Map.map (fn . This) (fromList [("both",2),("left-only",1)])
    -- fn (This 1)
    -- fn (This 2)
    -- Map.map (fn . That) (fromList [("both",4),("right-only",3)])
    -- fn (That 3)
    -- fn (That 4)
    -- True
    noisyMap :: (Show k, Show a) => String -> (a -> b) -> Map k a -> Map k b
    noisyMap showf f = noisyApply (printf "Map.map %s" showf) (Map.map f)

This time, as expected, `fn` gets call on both `These 2 4` and also needlessly on `This 2` and `That 4`.

**Comment by phadej** [comment]
There is stalled PR #51, if someone is willing to pull out the `Map` and `IntMap` changes and make new PR, I'll happily review and merge it
